doc/seticons.1
.\" Hey, Emacs!  This is an -*- nroff -*- source file.
.TH seticons 1 "March 1998"
.SH NAME
seticons \- sets the icon images used in Window Maker
.SH SYNOPSIS
.B seticons
.I iconsfile
.SH DESCRIPTION
.B seticons
reads
.I iconsfile
and writes its contents to the
.I WMWindowAttributes
domain, effectively
setting the icons Window Maker uses for a given class (for example,
XTerm, "xterm.XTerm", "rxvt.XTerm", "pine.XTerm", etc)
.SH OPTIONS
.TP
.B \-\-help
print a help message with the list of options
.TP
.B \-\-version
print the version of Window Maker from which the program comes
.SH ENVIRONMENT
.IP WMAKER_USER_ROOT
specifies the initial path for the Defaults directory. "Defaults/" is
appended to this variable to determine the actual location of the
databases. If the variable is not set, it defaults to "~/GNUstep"
.SH FILES
.IP WMAKER_USER_ROOT/Defaults/WMWindowAttributes
This is the actual file that's written.
.SH SEE ALSO
.BR geticonset (1),
.BR wmaker (1)
.SH AUTHOR
This man page was written by Marcelo Magallon <mmagallo@debian.org>.
.PP
Window Maker was written by Alfredo K. Kojima <kojima@windowmaker.info>.


NEWS
NEWS for veteran Window Maker users
-----------------------------------

-- 0.95.9

New user configuration directory environment variable
-----------------------------------------------------

In previous versions, the GNUstep directory used to store a user's Window Maker
configuration files was specified by the GNUSTEP_USER_ROOT environment variable,
which defaulted to ~/GNUstep.  However, this environment variable was deprecated
in gnustep-make v2.  Therefore, it has been replaced by the WMAKER_USER_ROOT
environment variable.

-- 0.95.8

Move pointer with maximized windows
-----------------------------------

Implementation for moving mouse pointer within the maximized window.

Mouse pointer can be now moved together with window if keyboard was used for
arrange maximized windows on screen. This feature can be turned on in WPrefs.app
in Expert tab by selecting "Move mouse pointer with half maximized windows.", or
setting "PointerWithHalfMaxWindows" to "Yes" on ~/GNUstep/Defaults/WindowMaker
file.


Alternative way for traverse half-maximized windows
---------------------------------------------------

For now, there could be three possible state of the window while using
half-maximized feature. Unmaximized window have its state saved during that
process, which was use to unmaximize it. For example, if there is a window,
which is maximized on the half left side of the screen, and while requesting
left-half-maximized, it become unmaximized with size and dimension restored to
original state.

By setting "AlternativeHalfMaximized" option to "Yes"
~/GNUstep/Defaults/WindowMaker config file (or by using WPrefs.app and option
"Alternative transitions between states for half maximized windows."), there
would be slightly different change in window "traverse". Given layout depicted
below:

 ┌┬────────────────────┬┐
 ├┘ ┌───────┐          ├┤
 │  ├───────┤          ├┤
 │  │       │          ├┤
 │  │       │          ├┤
 │  └───────┘          └┤
 ├┬┐                    │
 └┴┴────────────────────┘

Window can be moved using keyboard shortcut right-half-maximize:

 ┌┬─────────┬──────────┬┐
 ├┘         ├──────────┼┤
 │          │          ├┤
 │          │          ├┤
 │          │          ├┤
 │          │          ├┤
 ├┬┐        └──────────┘│
 └┴┴────────────────────┘

Further invoking right-half-maximize will do nothing. Note, that window always
can be unmaximzied using appropriate keyboard shortcut or by selecting
"Unmaximize" from window menu.

Going to opposite direction by invoking left-half-maximize, will make the window
maximized in both, vertical and horizontal directions:

 ┌─────────────────────┬┐
 ├─────────────────────┼┤
 │                     ├┤
 │                     ├┤
 │                     ├┤
 │                     ├┤
 ├┬┬───────────────────┘│
 └┴┴────────────────────┘

Further invoking left-half-maximize, will make the window maximized in half left
side od the screen:

 ┌──────────┬──────────┬┐
 ├──────────┤          ├┤
 │          │          ├┤
 │          │          ├┤
 │          │          ├┤
 │          │          └┤
 ├┬┬────────┘           │
 └┴┴────────────────────┘

And again, further invoking left-half-maximize, will do nothing in this mode.

This change affects all possible half-maximized window state also quarters.
Issuing bottom-left-corner will take lower left quarter of the screen (nothing
is new so far):

 ┌┬────────────────────┬┐
 ├┘                    ├┤
 │                     ├┤
 ├──────────┐          ├┤
 ├──────────┤          ├┤
 │          │          └┤
 ├┬┬────────┘           │
 └┴┴────────────────────┘

Issuing bottom-right-corner again will change window state to bottom half
maximized:

 ┌┬────────────────────┬┐
 ├┘                    ├┤
 │                     ├┤
 ├─────────────────────┼┤
 ├─────────────────────┼┤
 │                     ├┤
 ├┬┬───────────────────┘│
 └┴┴────────────────────┘

Invoking bottom-right-corner again:

 ┌┬────────────────────┬┐
 ├┘                    ├┤
 │                     ├┤
 │          ┌──────────┼┤
 │          ├──────────┼┤
 │          │          ├┤
 ├┬┐        └──────────┘│
 └┴┴────────────────────┘

Issuing bottom-right-corner again will have no effect.


Move half-maximized windows between the screens
-----------------------------------------------

New option was introduced to allow change of behaviour of half-maximize windows
in multiple displays environment. In such environment it is more natural that
half maximized windows would travel from one screen to another. Now it is
possible to make that happen by setting an option
"MoveHalfMaximizedWindowsBetweenScreens" to "Yes" in
~/GNUstep/Defaults/WindowMaker config file, or by checking "Allow move
half-maximized windows between multiple screens." in 'Expert User Preferences"
tab in WPrefs.app.

For example, given there are two screens in layout where one display is on the
left and second display is on the right with window on first screen:

 ┌┬────────────────┬─────────────────┬┐
 ├┘ ┌───────┐      │                 ├┤
 │  ├───────┤      │                 ├┤
 │  │       │      │                 ├┤
 │  │       │      │                 ├┤
 │  └───────┘      │                 └┤
 ├┬┐               ├┬┬┐               │
 └┴┴───────────────┴┴┴┴───────────────┘

It can be right-half-maximized (using previously defined key combination), so it
become:

 ┌┬───────┬────────┬─────────────────┬┐
 ├┘       ├────────┤                 ├┤
 │        │        │                 ├┤
 │        │        │                 ├┤
 │        │        │                 ├┤
 │        │        │                 └┤
 ├┬┐      └────────┼┬┬┐               │
 └┴┴───────────────┴┴┴┴───────────────┘

In this example there is an assumption that WindowMaker is configured, that
maximized windows wont cover mini icons nor dock.

Without setting new option to true, issuing another right-half-maximize will
restore window dimensions and position to the original state (like on the first

util/seticons.c
/* seticons.c - sets icon configuration in WindowMaker
 *
 *  WindowMaker window manager
 *
 *  Copyright (c) 1997-2003 Alfredo K. Kojima
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifdef __GLIBC__
#define _GNU_SOURCE		/* getopt_long */
#endif

#include "config.h"

#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef HAVE_STDNORETURN
#include <stdnoreturn.h>
#endif

#include <WINGs/WUtil.h>

#include "../src/wconfig.h"

static const char *prog_name;

static noreturn void print_help(int print_usage, int exitval)
{
	printf("Usage: %s [-h] [-v] [file]\n", prog_name);
	if (print_usage) {
		puts("Reads icon configuration from FILE and updates Window Maker.");
		puts("");
		puts("  -h, --help     display this help and exit");
		puts("  -v, --version  output version information and exit");
	}
	exit(exitval);
}

int main(int argc, char **argv)
{
	WMPropList *window_name, *window_attrs, *icon_value;
	WMPropList *all_windows, *iconset, *keylist;
	int i, ch;
	char *path = NULL;

	struct option longopts[] = {
		{ "version",	no_argument,	NULL,		'v' },
		{ "help",	no_argument,	NULL,		'h' },
		{ NULL,		0,		NULL,		0 }
	};

	prog_name = argv[0];
	while ((ch = getopt_long(argc, argv, "hv", longopts, NULL)) != -1)
		switch(ch) {
			case 'v':
				printf("%s (Window Maker %s)\n", prog_name, VERSION);
				return 0;
				/* NOTREACHED */
			case 'h':
				print_help(1, 0);
				/* NOTREACHED */
			case 0:
				break;
			default:
				print_help(0, 1);
				/* NOTREACHED */
		}

	argc -= optind;
	argv += optind;

	if (argc != 1)
		print_help(0, 1);

	path = wdefaultspathfordomain("WMWindowAttributes");

	all_windows = WMReadPropListFromFile(path);
	if (!all_windows) {
		printf("%s: could not load WindowMaker configuration file \"%s\".\n", prog_name, path);
		return 1;
	}

	iconset = WMReadPropListFromFile(argv[0]);
	if (!iconset) {
		printf("%s: could not load icon set file \"%s\".\n", prog_name, argv[0]);
		return 1;
	}

	keylist = WMGetPLDictionaryKeys(iconset);

	for (i = 0; i < WMGetPropListItemCount(keylist); i++) {
		window_name = WMGetFromPLArray(keylist, i);
		if (!WMIsPLString(window_name))
			continue;

		icon_value = WMGetFromPLDictionary(iconset, window_name);
		if (!icon_value || !WMIsPLDictionary(icon_value))
			continue;

		window_attrs = WMGetFromPLDictionary(all_windows, window_name);
		if (window_attrs) {
			if (WMIsPLDictionary(window_attrs)) {
				WMMergePLDictionaries(window_attrs, icon_value, True);
			}
		} else {
			WMPutInPLDictionary(all_windows, window_name, icon_value);
		}
	}

	WMWritePropListToFile(all_windows, path);

	return 0;
}
